include/irr/asset/IVirtualTexture.h

```diff
@@ -0,0 +1,750 @@
+            uint32_t tilesPerDim_log2;
+            uint32_t layerCount;
+        };
+
+        //_format implies format class and also is the format image is created with
+        IVTResidentStorage(uint32_t _layers, uint32_t _tilesPerDim) :
+            image(nullptr),//initialized in derived class's constructor
+            m_alctrReservedSpace(reinterpret_cast<uint8_t*>(_IRR_ALIGNED_MALLOC(phys_pg_addr_alctr_t::reserved_size(1u, _layers*_tilesPerDim*_tilesPerDim, 1u), _IRR_SIMD_ALIGNMENT))),
+            tileAlctr(m_alctrReservedSpace, 0u, 0u, 1u, _layers*_tilesPerDim*_tilesPerDim, 1u),
+            m_addr_layerShift(core::findLSB(_tilesPerDim)<<1),
+            m_addr_xMask((1u<<(m_addr_layerShift>>1))-1u)
+        {
+        }
+        //TODO: this should also copy address allocator state, but from what i see our address allocators doesnt have copy ctors
+        IVTResidentStorage(core::smart_refctd_ptr<image_t>&& _image, const core::vector<uint16_t>& _assignedLayers, uint32_t _layerShift, uint32_t _xmask) :
+            image(std::move(_image)),
+            m_assignedPageTableLayers(_assignedLayers),
+            m_addr_layerShift(_layerShift),
+            m_addr_xMask(_xmask)
+        {
+
+        }
+
+        core::smart_refctd_ptr<image_view_t> createView(E_FORMAT _format) const
+        {
+            auto found = m_viewsCache.find(_format);
+            if (found!=m_viewsCache.end())
+                return found->second;
+
+            typename image_view_t::SCreationParams params;
+            params.flags = static_cast<IImageView<image_t>::E_CREATE_FLAGS>(0);
+            params.format = _format;
+            params.subresourceRange.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+            params.subresourceRange.baseArrayLayer = 0u;
+            params.subresourceRange.layerCount = image->getCreationParameters().arrayLayers;
+            params.subresourceRange.baseMipLevel = 0u;
+            params.subresourceRange.levelCount = image->getCreationParameters().mipLevels;
+            params.image = core::smart_refctd_ptr<image_t>(image);
+            params.viewType = asset::IImageView<image_t>::ET_2D_ARRAY;
+
+            return m_viewsCache.insert({_format,createView_internal(std::move(params))}).first->second;
+        }
+        auto getPageTableLayers() const
+        {
+            return std::make_pair(m_assignedPageTableLayers.begin(),m_assignedPageTableLayers.end());
+        }
+        void addPageTableLayer(uint16_t _layer)
+        {
+            auto rng = getPageTableLayers();
+            auto it = std::lower_bound(rng.first,rng.second,_layer);
+            m_assignedPageTableLayers.insert(it, _layer);
+        }
+        void removeLayerAssignment(uint16_t _layer)
+        {
+            auto rng = getPageTableLayers();
+            auto it = std::lower_bound(rng.first, rng.second, _layer);
+            if (it!=rng.second && *it==_layer)
+                m_assignedPageTableLayers.erase(it);
+        }
```

Comment:
these functions shouldn't be public, make the VT classes friends or something

Reply:
anyway there's not much point keeping the mapping of format class->layer here (or at all)